pallene/util.lua

```diff
@@ -107,6 +107,26 @@ function util.outputs_of_execute(cmd)
     return ok, err, out_content, err_content
 end
 
+-- performs a deep copy of the table 'obj'
+
```

Comment:
This function is making my spider sense tingle. Smells fishy.
Did you get this from somewhere?

Reply:
I had this function saved for a few years (using across projects) but indeed I copied from somewhere in the web. I also have an extended version that handles metatables correctly. 

A quick search retrieved it being an answer from so https://stackoverflow.com/questions/640642/how-do-you-copy-a-lua-table-by-value , apparently it came from a gist

Reply:
Segue uma sugestão minha para deixar o código um pouco mais legível:

1. Se transformarmos o `seen` em um upvalue podemos nos livrar dos `if seen` e `or seen`
2. Achei o truque do `for k, v in next, obj` difícil de entender. Que tal usar um while?
3. Acho que dá pra melhorar os comentários.

Ah, e se puder adicionar uns casos de teste no util_spec seria show de bola. ⚽

```lua
-- Deep copy a Lua table or object.
-- Based on https://gist.github.com/tylerneylon/81333721109155b2d244
function deep_copy(root_obj)

    -- We memoize the tables to avoid looping
    -- forever if there are circular references.
    local memo = {}

    local function copy(obj)
        if type(obj) ~= 'table' then
            return obj
        else
            if memo[obj] then return memo[obj] end

            local res = {}
            memo[obj] = res

            -- Be careful with the metamethods:
            -- 1) Use `next` to iterate over the source table. It acts like "rawget".
            -- 2) Only set the destination's metatable after we finish setting its keys.
            local k, v = next(obj)
            while k do
                res[copy(k)] = copy(v)
                k, v = next(obj, k)
            end
            return setmetatable(res, getmetatable(obj))
        end
    end

    return copy(root_obj)
end

```